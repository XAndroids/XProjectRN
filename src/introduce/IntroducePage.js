import React, { Component } from 'react'
import { View, Text, Image, StyleSheet, Platform } from 'react-native';
import PropTypes from 'prop-types';
import { getWindowsWidth, getWindowsHeight } from '../utils/Devices';

/**
 * 定义了Introduce页面ViewPager子页面的内容
 */
export default class IntroducePage extends Component {

    render() {
        return (
            <View style={[styles.page, {
                backgroundColor: this.props.backgroundColor,
                width: getWindowsWidth(),
                height: getWindowsHeight()
            }]}>
                {this.renderText(this.props.title, 28)}
                <Image style={styles.image} source={require('../../img/introduce_icon1.png')} />
                {this.renderText(this.props.content, 16)}
            </View>
        )
    }

    /**
   * 渲染文案，iOS平台Text无法通过textAlignVertical实现字体居中，则使用View容器包裹实现
   * @param {显示的文案} text
   * @param {显示的文案字体大小} fontSize
   */
    renderText(text, fontSize) {
        if (Platform.OS === 'ios') {
            return (
                <View style={{ flex: 1, justifyContent: 'center' }}>
                    <Text style={[styles.text, { fontSize: fontSize }]}>{text}</Text>
                </View>);
        } else {
            return (
                <Text style={[styles.text, { fontSize: fontSize }]}>{text}</Text>
            );
        }
    }
}

IntroducePage.propTypes = {
    title: PropTypes.string.isRequired,
    image: PropTypes.string.isRequired,
    content: PropTypes.string.isRequired,
    backgroundColor: PropTypes.string.isRequired,
};

const styles = StyleSheet.create({
    page: {
        flex: 1,
        alignItems: 'center'
    },
    text: {
        color: 'white',
        ...Platform.select({
            android: {
                flex: 1,
                textAlign: 'center',
                textAlignVertical: 'center',
            },
        }),
    },
    image: {
        flex: 1,
    },
});

