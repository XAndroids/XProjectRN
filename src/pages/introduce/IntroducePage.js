import React, {Component} from 'react'
import {View, Text, Image, StyleSheet, Platform} from 'react-native'
import PropTypes from 'prop-types'
import {getWindowsWidth, getWindowsHeight} from '../../utils'

/**
 * 定义了Introduce页面ViewPager子页面的内容
 */
export default class IntroducePage extends Component {

    render() {
        const {backgroundColor, title, content, image} = this.props;

        return (
            <View style={[styles.page, {
                backgroundColor: backgroundColor,
                width: getWindowsWidth(),
                height: getWindowsHeight()
            }]}>
                {this._renderText(title, 28, 'center', 'center')}
                <Image style={{flex: 1}} source={image}/>
                {this._renderText(content, 16, 'flex-end', 'bottom')}
                <View style={{
                    backgroundColor: 'rgba(0, 0, 0, 0.1)',
                    height: 1,
                    width: getWindowsWidth(),
                    marginBottom: 50,
                    marginTop: 20
                }}/>
            </View>
        )
    }

    _renderText(text, fontSize, justifyContent, textAlignVertical) {
        if (Platform.OS === 'ios') {
            return (
                //渲染文案，iOS平台Text无法通过textAlignVertical实现字体居中，则使用View容器包裹实现
                <View style={{flex: 1, justifyContent: justifyContent}}>
                    <Text style={[styles.text, {fontSize: fontSize}]}>{text}</Text>
                </View>);
        } else {
            return (
                <Text style={[styles.text, {
                    fontSize: fontSize,
                    textAlignVertical: textAlignVertical
                }]}>{text}</Text>
            );
        }
    }
}

IntroducePage.propTypes = {
    title: PropTypes.string.isRequired,
    image: PropTypes.number.isRequired,
    content: PropTypes.string.isRequired,
    backgroundColor: PropTypes.string.isRequired,
};

const styles = StyleSheet.create({
    page: {
        flex: 1,
        alignItems: 'center'
    },
    text: {
        color: 'white',
        textAlign: 'center',
        ...Platform.select({
            android: {
                flex: 1
            },
        }),
    }
});

