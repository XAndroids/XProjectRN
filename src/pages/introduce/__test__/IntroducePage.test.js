import React from "react";
import {Image,View} from "react-native"
import {render} from "react-native-testing-library";
import IntroducePage from "../IntroducePage";

describe("IntroducePage test",()=>{
    const page = {
        title: 'Welcome',
        image: require('../../../../img/introduce_icon1.png'),
        content: 'Thank you for installing Paperboy.What makes Paperboy different form other news reader apps is its simplicity and elegant design.',
        backgroundColor: '#01A3AE',
    }

    beforeAll(()=>{
        //测试android Platfor逻辑
        //jest.requireActual(moduleName):返回一个实际的对象替代mock
        //参考：https://jestjs.io/docs/en/jest-object#jestrequireactualmodulename
        jest.mock("Platform",()=>{
            const Platform = require.requireActual('Platform');
            Platform.OS = 'android';
            return Platform;
        })
    })

    describe("IntroducePage property test",()=>{
        it("title property test",()=>{
            const{queryByText} = render(<IntroducePage title={page.title} image={page.image} content={page.content}
                                                       backgroundColor={page.backgroundColor}/>)
            //title文案显示，并且样式正常
            const titleText = queryByText(page.title)
            expect(titleText).not.toBeNull()
            expect(titleText.props.style).toEqual( [{"color": "white", "textAlign": "center"}, {"fontSize": 28, "textAlignVertical": "center"}])
        })

        it("image property test",()=>{
            const{queryByType} = render(<IntroducePage title={page.title} image={page.image} content={page.content}
                                                                   backgroundColor={page.backgroundColor}/>)
            //图片显示，并且样式和source正常
            const pageImage = queryByType(Image)
            expect(pageImage).not.toBeNull()
            expect(pageImage.props.style).toEqual({ flex: 1 })
            expect(pageImage.props.source.testUri).toBe('../../../img/introduce_icon1.png')
        })

        it("content property test",()=>{
            const{queryByText} = render(<IntroducePage title={page.title} image={page.image} content={page.content}
                                                       backgroundColor={page.backgroundColor}/>)
            //内容文案显示，并且样式正常
            const contentText = queryByText(page.content)
            expect(contentText).not.toBeNull()
            expect(contentText.props.style).toEqual( [{ color: 'white', textAlign: 'center' },
                { fontSize: 16, textAlignVertical: 'bottom' }])
        })

        it("backgroundColor property test",()=>{
            const{queryAllByType} = render(<IntroducePage title={page.title} image={page.image} content={page.content}
                                                          backgroundColor={page.backgroundColor}/>)

            //页面背景颜色正常
            const rootView = queryAllByType(View)[0]
            expect(rootView).not.toBeNull()
            expect(rootView.props.style[1].backgroundColor).toBe(page.backgroundColor)
        })
    })

    describe("IntroducePage Snapshot test", () => {
        it("IntroducePage Snapshot test", () => {
            const{toJSON} = render(<IntroducePage title={page.title} image={page.image} content={page.content}
                                                          backgroundColor={page.backgroundColor}/>)
            expect(toJSON()).toMatchSnapshot()
        })
    })

    describe("IntroducePage shallow test", () => {
        it("IntroducePage shallow test", () => {
            const{output} = render(<IntroducePage title={page.title} image={page.image} content={page.content}
                                                  backgroundColor={page.backgroundColor}/>)
            expect(output).toMatchSnapshot()
        })
    })
})
