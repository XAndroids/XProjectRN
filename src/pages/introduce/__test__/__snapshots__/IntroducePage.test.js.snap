// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IntroducePage test IntroducePage Snapshot test IntroducePage Snapshot test 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flex": 1,
      },
      Object {
        "backgroundColor": "#01A3AE",
        "height": 1334,
        "width": 750,
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "white",
          "textAlign": "center",
        },
        Object {
          "fontSize": 28,
          "textAlignVertical": "center",
        },
      ]
    }
  >
    Welcome
  </Text>
  <Image
    source={
      Object {
        "testUri": "../../../img/introduce_icon1.png",
      }
    }
    style={
      Object {
        "flex": 1,
      }
    }
  />
  <Text
    style={
      Array [
        Object {
          "color": "white",
          "textAlign": "center",
        },
        Object {
          "fontSize": 16,
          "textAlignVertical": "bottom",
        },
      ]
    }
  >
    Thank you for installing Paperboy.What makes Paperboy different form other news reader apps is its simplicity and elegant design.
  </Text>
  <View
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.1)",
        "height": 1,
        "marginBottom": 50,
        "marginTop": 20,
        "width": 750,
      }
    }
  />
</View>
`;

exports[`IntroducePage test IntroducePage shallow test IntroducePage shallow test 1`] = `undefined`;
